File: android-aop-plugin/src/main/kotlin/com/flyjingfish/android_aop_plugin/scanner_visitor/AnnotationMethodScanner.java
package com.flyjingfish.android_aop_plugin.scanner_visitor;

import com.flyjingfish.android_aop_plugin.beans.AopMatchCut;
import com.flyjingfish.android_aop_plugin.beans.LambdaMethod;
import com.flyjingfish.android_aop_plugin.beans.MatchMethodInfo;
import com.flyjingfish.android_aop_plugin.beans.MethodRecord;
import com.flyjingfish.android_aop_plugin.utils.ClassPoolUtils;
import com.flyjingfish.android_aop_plugin.utils.Utils;
import com.flyjingfish.android_aop_plugin.utils.WovenInfoUtils;

import org.objectweb.asm.AnnotationVisitor;
import org.objectweb.asm.Handle;
import org.objectweb.asm.MethodVisitor;
import org.objectweb.asm.Opcodes;
import org.objectweb.asm.Type;
import org.objectweb.asm.tree.AbstractInsnNode;
import org.objectweb.asm.tree.ClassNode;
import org.objectweb.asm.tree.InvokeDynamicInsnNode;
import org.objectweb.asm.tree.MethodNode;
import org.slf4j.Logger;

import java.io.ByteArrayInputStream;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;

import javassist.ClassPool;
import javassist.CtClass;
import javassist.CtMethod;
import javassist.NotFoundException;
import javassist.bytecode.CodeAttribute;
import javassist.bytecode.MethodInfo;

public class AnnotationMethodScanner extends ClassNode {
    Logger logger;
    private final OnCallBackMethod onCallBackMethod;
//    private final byte[] classByte;
    private final List<AopMatchCut> aopMatchCuts = new ArrayList<>();
    private final List<MethodRecord> cacheMethodRecords = new ArrayList<>();
    private String className;

    public AnnotationMethodScanner(Logger logger,OnCallBackMethod onCallBackMethod) {
        super(Opcodes.ASM9);
        this.logger = logger;
        this.onCallBackMethod = onCallBackMethod;
//        this.classByte = classByte;
    }

    @Override
    public void visit(int version, int access, String name, String signature, String superName, String[] interfaces) {
        className = name;
//        logger.error("className="+className+",superName="+superName+",interfaces="+ Arrays.asList(interfaces));
        WovenInfoUtils.INSTANCE.getAopMatchCuts().forEach((key, aopMatchCut) -> {
            if (!AopMatchCut.MatchType.SELF.name().equals(aopMatchCut.getMatchType())){
                String[] excludeClazz = aopMatchCut.getExcludeClass();
                boolean exclude = false;
                boolean isDirectExtends = false;
                if (excludeClazz != null){
                    String clsName = Utils.INSTANCE.slashToDotClassName(className);
                    for (String clazz : excludeClazz) {
                        if (clsName.equals(Utils.INSTANCE.slashToDotClassName(clazz))){
                            exclude = true;
                            break;
                        }
                    }
                }

                if (!exclude){
                    boolean isImplementsInterface = false;
                    if (AopMatchCut.MatchType.EXTENDS.name().equals(aopMatchCut.getMatchType())){
                        if (interfaces != null) {
                            for (String anInterface : interfaces) {
                                String inter = Utils.INSTANCE.slashToDotClassName(anInterface);
                                if (inter.equals(Utils.INSTANCE.slashToDotClassName(aopMatchCut.getBaseClassName()))) {
                                    isImplementsInterface = true;
                                    break;
                                }
                            }
                        }
                    }

                    if (isImplementsInterface
                            || (Utils.INSTANCE.slashToDotClassName(aopMatchCut.getBaseClassName()).equals(Utils.INSTANCE.slashToDotClassName(superName)))
                    ) {
                        isDirectExtends = true;
//                        this.isDescendantClass = true;
//                        AnnotationMethodScanner.this.aopMatchCuts.add(aopMatchCut);
                    }
                    //isDirectExtends 为true 说明是直接继承
                    if (AopMatchCut.MatchType.DIRECT_EXTENDS.name().equals(aopMatchCut.getMatchType())){
                        if (isDirectExtends){
                            AnnotationMethodScanner.this.aopMatchCuts.add(aopMatchCut);
                        }
                    }else if (AopMatchCut.MatchType.LEAF_EXTENDS.name().equals(aopMatchCut.getMatchType())){
                        boolean isExtends = false;
                        if (isDirectExtends){
                            isExtends = true;

                        }else {
                            String clsName = Utils.INSTANCE.slashToDotClassName(className);
                            String parentClsName = aopMatchCut.getBaseClassName();
                            if (!clsName.equals(Utils.INSTANCE.slashToDotClassName(parentClsName))) {
                                try {

                                    isExtends = Utils.INSTANCE.isInstanceof(clsName,Utils.INSTANCE.slashToDotClassName(parentClsName));
//                        logger.error("isInstanceof="+isInstanceof);
                                } catch (NotFoundException e) {
                                    e.printStackTrace();
                                }

                            }
                        }
                        if (isExtends && WovenInfoUtils.INSTANCE.isLeaf(className)){
                            AnnotationMethodScanner.this.aopMatchCuts.add(aopMatchCut);
                        }
                    }else {
                        if (isDirectExtends){
                            AnnotationMethodScanner.this.aopMatchCuts.add(aopMatchCut);
                        }else {
                            String clsName = Utils.INSTANCE.slashToDotClassName(className);
                            String parentClsName = aopMatchCut.getBaseClassName();
                            if (!clsName.equals(Utils.INSTANCE.slashToDotClassName(parentClsName))) {
                                try {

                                    boolean isInstanceof = Utils.INSTANCE.isInstanceof(clsName,Utils.INSTANCE.slashToDotClassName(parentClsName));
//                        logger.error("isInstanceof="+isInstanceof);
                                    if (isInstanceof){
                                        AnnotationMethodScanner.this.aopMatchCuts.add(aopMatchCut);
                                    }

                                } catch (NotFoundException e) {
                                    e.printStackTrace();
                                }

                            }
                        }
                    }


                }
            }
            if ((AopMatchCut.MatchType.SELF.name().equals(aopMatchCut.getMatchType()) && Utils.INSTANCE.slashToDotClassName(aopMatchCut.getBaseClassName()).equals(Utils.INSTANCE.slashToDotClassName(name)))) {
                AnnotationMethodScanner.this.aopMatchCuts.add(aopMatchCut);
            }
        });
        super.visit(version, access, name, signature, superName, interfaces);

    }

    @Override
    public AnnotationVisitor visitAnnotation(String descriptor, boolean visible) {
        return super.visitAnnotation(descriptor, visible);
    }

    class MyMethodVisitor extends MethodNode {
        MethodRecord methodName;

//        MyMethodVisitor(MethodRecord methodName) {
//            super(Opcodes.ASM8);
//            this.methodName = methodName;
//        }

        public MyMethodVisitor(
                final int access,
                final String name,
                final String descriptor,
                final String signature,
                final String[] exceptions,MethodRecord methodName) {
            super(/* latest api = */ Opcodes.ASM9, access, name, descriptor, signature, exceptions);
            this.methodName = methodName;
        }

        @Override
        public AnnotationVisitor visitAnnotation(String descriptor, boolean visible) {
//            logger.error("AnnotationMethodScanner MyMethodVisitor type: " + descriptor);
            if (WovenInfoUtils.INSTANCE.isContainAnno(descriptor)) {
                boolean isBack = true;
                StringBuilder paramStr = new StringBuilder();
                try {
                    ClassPool classPool = ClassPoolUtils.INSTANCE.getClassPool();
//                    InputStream byteArrayInputStream = new ByteArrayInputStream(classByte);
//                    CtClass ctClass = classPool.makeClass(byteArrayInputStream);
                    String clsName = Utils.INSTANCE.slashToDot(className);
                    CtClass ctClass =  classPool.get(clsName);
                    CtMethod ctMethod = WovenIntoCode.INSTANCE.getCtMethod(ctClass, methodName.getMethodName(), methodName.getDescriptor());
                    MethodInfo methodInfo = ctMethod.getMethodInfo();
                    CodeAttribute codeAttribute = methodInfo.getCodeAttribute();
                    if (codeAttribute == null) {
                        isBack = false;
                    }else {
                        CtClass[] ctClasses = ctMethod.getParameterTypes();
                        paramStr.append("(");
                        int length = ctClasses.length;
                        for (int i = 0; i < length; i++) {
                            paramStr.append(ctClasses[i].getName());
                            if (i != length - 1) {
                                paramStr.append(",");
                            }
                        }
                        paramStr.append(")");

                        CtClass returnCtClass = ctMethod.getReturnType();
                        String returnType = returnCtClass.getName();
                    }
                } catch (Exception ignored) {
                }
                if (onCallBackMethod != null && isBack) {
                    onCallBackMethod.onBackName(methodName);

                }
            }
            return super.visitAnnotation(descriptor, visible);
        }
    }

    @Override
    public MethodVisitor visitMethod(int access, String name, String descriptor,
                                     String signature, String[] exceptions) {

        if (aopMatchCuts.size() > 0) {
            for (AopMatchCut aopMatchCut : aopMatchCuts) {
                for (String methodName : aopMatchCut.getMethodNames()) {
                    MatchMethodInfo matchMethodInfo = Utils.INSTANCE.getMethodInfo(methodName);
                    if (matchMethodInfo != null && name.equals(matchMethodInfo.getName())) {
                        boolean isBack = true;
                        try {
                            ClassPool classPool = ClassPoolUtils.INSTANCE.getClassPool();
                            String clsName = Utils.INSTANCE.slashToDot(className);
                            CtClass ctClass =  classPool.get(clsName);
//                            InputStream byteArrayInputStream = new ByteArrayInputStream(classByte);
//                            CtClass ctClass = classPool.makeClass(byteArrayInputStream);
                            CtMethod ctMethod = WovenIntoCode.INSTANCE.getCtMethod(ctClass, name, descriptor);
                            if (matchMethodInfo.getParamTypes() != null) {
                                CtClass[] ctClasses = ctMethod.getParameterTypes();
                                StringBuilder paramStr = new StringBuilder();
                                paramStr.append("(");
                                int length = ctClasses.length;
                                for (int i = 0; i < length; i++) {
                                    paramStr.append(ctClasses[i].getName());
                                    if (i != length - 1) {
                                        paramStr.append(",");
                                    }
                                }
                                paramStr.append(")");
                                //有设置参数类型这一项
                                if (!paramStr.toString().equals(matchMethodInfo.getParamTypes())) {
                                    isBack = false;
                                }
                            }
                            if (matchMethodInfo.getReturnType() != null) {
                                CtClass returnCtClass = ctMethod.getReturnType();
                                String returnType = returnCtClass.getName();
                                //有设置返回类型这一项
                                if (!returnType.equals(matchMethodInfo.getReturnType())) {
                                    isBack = false;
                                }
                            }
//                            logger.error("paramStr="+paramStr+",returnType="+returnType+",matchMethodInfo="+matchMethodInfo);
                            MethodInfo methodInfo = ctMethod.getMethodInfo();
                            CodeAttribute codeAttribute = methodInfo.getCodeAttribute();
                            if (codeAttribute == null) {
                                isBack = false;
                            }
                        } catch (Exception ignored) {
                        }
                        if (isBack) {
                            cacheMethodRecords.add(new MethodRecord(name, descriptor, aopMatchCut.getCutClassName()));
                        }
                        break;
                    }
                }
            }
        }
        MyMethodVisitor myMethodVisitor = new MyMethodVisitor(access,name,descriptor,signature,exceptions,new MethodRecord(name, descriptor, null));
        methods.add(myMethodVisitor);
        return myMethodVisitor;
    }
    List<LambdaMethod> lambdaMethodList = new ArrayList<>();
    @Override
    public void visitEnd() {
        super.visitEnd();

        this.methods.forEach(methodNode -> {
//            logger.error(className+"======methods.forEach=");
            for (AbstractInsnNode node : methodNode.instructions) {
                if (node instanceof InvokeDynamicInsnNode tmpNode) {
                    String desc = tmpNode.desc;
                    Type descType = Type.getType(desc);
                    Type samBaseType = descType.getReturnType();
                    //sam 接口名
                    String samBase = samBaseType.getDescriptor();
                    //sam 方法名
                    String samMethodName = tmpNode.name;
                    Object[] bsmArgs = tmpNode.bsmArgs;
                            Type samMethodType = (Type) bsmArgs[0];
                    Handle methodName = (Handle) bsmArgs[1];
                    //sam 实现方法实际参数描述符
//                            Type implMethodType = (Type) bsmArgs[2];

                    String lambdaName = methodName.getName();
//                    String bsmMethodNameAndDescriptor = samMethodName + samMethodType.getDescriptor();
                    String thisClassName = Utils.INSTANCE.slashToDot(samBase.substring(1).replaceAll(";","")).replaceAll("\\$", ".");
                    String originalClassName = Utils.INSTANCE.slashToDot(samBase.substring(1).replaceAll(";",""));
//                    logger.error("className="+className+",tmpNode.name="+tmpNode.name+",desc=" + desc + ",samBase=" + samBase + ",samMethodName="
//                            + samMethodName + ",methodName=" + lambdaName+ ",methodDesc=" + methodName.getDesc()+",thisClassName="+thisClassName+
//                            ",getDescriptor"+samMethodType.getDescriptor());
                    String lambdaDesc = methodName.getDesc();
                    String samMethodDesc = samMethodType.getDescriptor();
                    LambdaMethod lambdaMethod = new LambdaMethod(samMethodName,samMethodDesc,thisClassName,originalClassName,lambdaName,lambdaDesc);
                    lambdaMethodList.add(lambdaMethod);
                }

            }
            if (lambdaMethodList.size() > 0){
//            logger.error(className+"======lambdaMethodList="+lambdaMethodList);
                WovenInfoUtils.INSTANCE.getAopMatchCuts().forEach((key, aopMatchCut) -> {
//                if (className.contains("MainActivity")){
//                    logger.error("======aopMatchCut="+aopMatchCut+"=size"+WovenInfoUtils.INSTANCE.getAopMatchCuts().size());
//                }
                    if (!AopMatchCut.MatchType.SELF.name().equals(aopMatchCut.getMatchType()) && aopMatchCut.getMethodNames().length == 1){
                        for (LambdaMethod lambdaMethod : lambdaMethodList) {
                            boolean isDirectExtends = Utils.INSTANCE.slashToDotClassName(aopMatchCut.getBaseClassName()).equals(lambdaMethod.getThisClassName());
//                            logger.error("lambdaMethod="+lambdaMethod+",parent="+parent);
                            boolean isMatch = false;

                            if (AopMatchCut.MatchType.DIRECT_EXTENDS.name().equals(aopMatchCut.getMatchType())){
                                isMatch = isDirectExtends;
                            }else if (AopMatchCut.MatchType.EXTENDS.name().equals(aopMatchCut.getMatchType())
                                    ||AopMatchCut.MatchType.LEAF_EXTENDS.name().equals(aopMatchCut.getMatchType())){
                                isMatch = isDirectExtends;
                                if (!isMatch){
                                    String clsName = lambdaMethod.getThisClassName();
                                    String parentClsName = aopMatchCut.getBaseClassName();
                                    try {
                                        isMatch = Utils.INSTANCE.isInstanceof(clsName,Utils.INSTANCE.slashToDotClassName(parentClsName));
//                                        logger.error("className="+className+"lambdaMethod="+lambdaMethod+",isInstanceof="+parent);

                                    } catch (NotFoundException e) {
                                        e.printStackTrace();
                                    }
                                }
                            }

                            String name = lambdaMethod.getSamMethodName();
                            String descriptor = lambdaMethod.getSamMethodDesc();
                            String aopMatchCutMethodName = aopMatchCut.getMethodNames()[0];

                            MatchMethodInfo matchMethodInfo = Utils.INSTANCE.getMethodInfo(aopMatchCutMethodName);
                            if (isMatch && matchMethodInfo != null && name.equals(matchMethodInfo.getName())) {
                                boolean isBack = true;
                                try {
                                    ClassPool classPool = ClassPoolUtils.INSTANCE.getClassPool();
                                    CtMethod ctMethod = null;
                                    if (matchMethodInfo.getParamTypes() != null || matchMethodInfo.getReturnType() != null) {
                                        CtClass ctClass = classPool.get(lambdaMethod.getOriginalClassName());
                                        ctMethod = WovenIntoCode.INSTANCE.getCtMethod(ctClass, name, descriptor);
                                    }
                                    if (matchMethodInfo.getParamTypes() != null && ctMethod != null) {
                                        CtClass[] ctClasses = ctMethod.getParameterTypes();
                                        StringBuilder paramStr = new StringBuilder();
                                        paramStr.append("(");
                                        int length = ctClasses.length;
                                        for (int i = 0; i < length; i++) {
                                            paramStr.append(ctClasses[i].getName());
                                            if (i != length - 1) {
                                                paramStr.append(",");
                                            }
                                        }
                                        paramStr.append(")");
                                        //有设置参数类型这一项
                                        if (!paramStr.toString().equals(matchMethodInfo.getParamTypes())) {
                                            isBack = false;
                                        }
                                    }
                                    if (matchMethodInfo.getReturnType() != null && ctMethod != null) {
                                        CtClass returnCtClass = ctMethod.getReturnType();
                                        String returnType = returnCtClass.getName();
                                        //有设置返回类型这一项
                                        if (!returnType.equals(matchMethodInfo.getReturnType())) {
                                            isBack = false;
                                        }
                                    }
//                            logger.error("paramStr="+paramStr+",returnType="+returnType+",matchMethodInfo="+matchMethodInfo);
                                } catch (Exception ignored) {
//                                    ignored.printStackTrace();
                                }
                                if (isBack) {
                                    MethodRecord methodRecord = new MethodRecord(lambdaMethod.getLambdaName(), lambdaMethod.getLambdaDesc(), aopMatchCut.getCutClassName());
                                    if (onCallBackMethod != null) {
                                        onCallBackMethod.onBackName(methodRecord);
                                    }
                                }
                            }


                        }
                    }
                });
            }
        });


        if (aopMatchCuts.size() > 0) {
            for (MethodRecord cacheMethodRecord : cacheMethodRecords) {
                if (onCallBackMethod != null) {
                    onCallBackMethod.onBackName(cacheMethodRecord);
                }
            }
        }



    }



    public interface OnCallBackMethod {
        void onBackName(MethodRecord methodRecord);
    }
}

File: android-aop-plugin/src/main/kotlin/com/flyjingfish/android_aop_plugin/scanner_visitor/AnnotationMethodScanner.kt
package com.flyjingfish.android_aop_plugin.scanner_visitor


import com.flyjingfish.android_aop_plugin.beans.AopMatchCut
import com.flyjingfish.android_aop_plugin.beans.CutMethodJson
import com.flyjingfish.android_aop_plugin.beans.LambdaMethod
import com.flyjingfish.android_aop_plugin.beans.MethodRecord
import com.flyjingfish.android_aop_plugin.config.AndroidAopConfig.Companion.cutInfoJson
import com.flyjingfish.android_aop_plugin.scanner_visitor.WovenIntoCode.getCtMethod
import com.flyjingfish.android_aop_plugin.utils.ClassPoolUtils.classPool
import com.flyjingfish.android_aop_plugin.utils.InitConfig.putCutInfo
import com.flyjingfish.android_aop_plugin.utils.Utils.getMethodInfo
import com.flyjingfish.android_aop_plugin.utils.Utils.isInstanceof
import com.flyjingfish.android_aop_plugin.utils.Utils.slashToDot
import com.flyjingfish.android_aop_plugin.utils.Utils.slashToDotClassName
import com.flyjingfish.android_aop_plugin.utils.WovenInfoUtils
import com.flyjingfish.android_aop_plugin.utils.WovenInfoUtils.getAnnoInfo
import com.flyjingfish.android_aop_plugin.utils.WovenInfoUtils.isContainAnno
import com.flyjingfish.android_aop_plugin.utils.WovenInfoUtils.isLeaf
import javassist.CtMethod
import org.objectweb.asm.AnnotationVisitor
import org.objectweb.asm.Handle
import org.objectweb.asm.MethodVisitor
import org.objectweb.asm.Opcodes
import org.objectweb.asm.Type
import org.objectweb.asm.tree.ClassNode
import org.objectweb.asm.tree.InvokeDynamicInsnNode
import org.objectweb.asm.tree.MethodNode
import org.slf4j.Logger


class AnnotationMethodScanner(val logger: Logger, val onCallBackMethod: OnCallBackMethod) :
    ClassNode(Opcodes.ASM9) {
    //    private final byte[] classByte;
    private val aopMatchCuts = mutableListOf<AopMatchCut>();

    //    private final List<MethodRecord> cacheMethodRecords = new ArrayList<>();
    lateinit var className: String


    override fun visit(
        version: Int,
        access: Int,
        name: String,
        signature: String?,
        superName: String?,
        interfaces: Array<String>?
    ) {
        className = name
        //        logger.error("className="+className+",superName="+superName+",interfaces="+ Arrays.asList(interfaces));
        WovenInfoUtils.aopMatchCuts.forEach { (_: String?, aopMatchCut: AopMatchCut) ->
            if (AopMatchCut.MatchType.SELF.name != aopMatchCut.matchType) {
                val excludeClazz = aopMatchCut.excludeClass
                var exclude = false
                var isDirectExtends = false
                if (excludeClazz != null) {
                    val clsName = slashToDotClassName(className)
                    for (clazz in excludeClazz) {
                        if (clsName == slashToDotClassName(clazz)) {
                            exclude = true
                            break
                        }
                    }
                }
                if (!exclude) {
                    var isImplementsInterface = false
                    if (AopMatchCut.MatchType.EXTENDS.name == aopMatchCut.matchType) {
                        if (interfaces != null) {
                            for (anInterface in interfaces) {
                                val inter = slashToDotClassName(anInterface)
                                if (inter == slashToDotClassName(aopMatchCut.baseClassName)) {
                                    isImplementsInterface = true
                                    break
                                }
                            }
                        }
                    }
                    if (isImplementsInterface || slashToDotClassName(aopMatchCut.baseClassName) == slashToDotClassName(
                            superName!!
                        )
                    ) {
                        isDirectExtends = true
                    }
                    //isDirectExtends 为true 说明是直接继承
                    if (AopMatchCut.MatchType.DIRECT_EXTENDS.name == aopMatchCut.matchType) {
                        if (isDirectExtends) {
                            aopMatchCuts.add(aopMatchCut)
                        }
                    } else if (AopMatchCut.MatchType.LEAF_EXTENDS.name == aopMatchCut.matchType) {
                        var isExtends = false
                        if (isDirectExtends) {
                            isExtends = true
                        } else {
                            val clsName = slashToDotClassName(className)
                            val parentClsName = aopMatchCut.baseClassName
                            if (clsName != slashToDotClassName(parentClsName)) {
                                isExtends = isInstanceof(clsName, slashToDotClassName(parentClsName))
                            }
                        }
                        if (isExtends && isLeaf(className)) {
                            aopMatchCuts.add(aopMatchCut)
                        }
                    } else {
                        if (isDirectExtends) {
                            aopMatchCuts.add(aopMatchCut)
                        } else {
                            val clsName = slashToDotClassName(className)
                            val parentClsName = aopMatchCut.baseClassName
                            if (clsName != slashToDotClassName(parentClsName)) {
                                val isInstanceof = isInstanceof(clsName, slashToDotClassName(parentClsName))
                                if (isInstanceof) {
                                    aopMatchCuts.add(aopMatchCut)
                                }
                            }
                        }
                    }
                }
            }
            if (AopMatchCut.MatchType.SELF.name == aopMatchCut.matchType && slashToDotClassName(
                    aopMatchCut.baseClassName
                ) == slashToDotClassName(name)
            ) {
                aopMatchCuts.add(aopMatchCut)
            }
        }
        super.visit(version, access, name, signature, superName, interfaces)
    }

    override fun visitAnnotation(descriptor: String?, visible: Boolean): AnnotationVisitor? {
        return super.visitAnnotation(descriptor, visible)
    }

    open inner class MyMethodVisitor
        (
        access: Int,
        name: String,
        descriptor: String?,
        signature: String?,
        exceptions: Array<String?>?, var methodName: MethodRecord
    ) : MethodNode(
        Opcodes.ASM9, access,
        name,
        descriptor,
        signature,
        exceptions
    ) {
        override fun visitAnnotation(descriptor: String, visible: Boolean): AnnotationVisitor? {
//            logger.error("AnnotationMethodScanner MyMethodVisitor type: " + descriptor);
            if (isContainAnno(descriptor)) {
                var isBack = true
                try {
                    val classPool = classPool
                    //                    InputStream byteArrayInputStream = new ByteArrayInputStream(classByte);
//                    CtClass ctClass = classPool.makeClass(byteArrayInputStream);
                    val clsName = slashToDot(className)
                    val ctClass = classPool!![clsName]
                    val ctMethod = getCtMethod(ctClass, methodName.methodName, methodName.descriptor)
                    val methodInfo = ctMethod!!.methodInfo
                    val codeAttribute = methodInfo.codeAttribute
                    if (codeAttribute == null) {
                        isBack = false
                    }
                } catch (ignored: Exception) {
                }
                if (isBack) {
                    onCallBackMethod.onBackName(methodName)
                    val aopMethodCut = getAnnoInfo(descriptor)
                    if (aopMethodCut != null && cutInfoJson) {
                        putCutInfo(
                            "注解切面", slashToDot(className), aopMethodCut.anno,
                            CutMethodJson(methodName.methodName, methodName.descriptor, false)
                        )
                    }
                }
            }
            return super.visitAnnotation(descriptor, visible)
        }
    }

    override fun visitMethod(
        access: Int, name: String, descriptor: String,
        signature: String?, exceptions: Array<String?>?
    ): MethodVisitor {
        if (aopMatchCuts.size > 0) {
            for (aopMatchCut in aopMatchCuts) {
                for (methodName in aopMatchCut.methodNames) {
                    val matchMethodInfo = getMethodInfo(methodName)
                    if (matchMethodInfo != null && name == matchMethodInfo.name) {
                        var isBack = true
                        try {
                            val classPool = classPool
                            val clsName = slashToDot(className)
                            val ctClass = classPool!![clsName]
                            //                            InputStream byteArrayInputStream = new ByteArrayInputStream(classByte);
//                            CtClass ctClass = classPool.makeClass(byteArrayInputStream);
                            val ctMethod = getCtMethod(ctClass, name, descriptor)
                            if (matchMethodInfo.paramTypes != null) {
                                val ctClasses = ctMethod!!.parameterTypes
                                val paramStr = java.lang.StringBuilder()
                                paramStr.append("(")
                                val length = ctClasses.size
                                for (i in 0 until length) {
                                    paramStr.append(ctClasses[i].name)
                                    if (i != length - 1) {
                                        paramStr.append(",")
                                    }
                                }
                                paramStr.append(")")
                                //有设置参数类型这一项
                                if (paramStr.toString() != matchMethodInfo.paramTypes) {
                                    isBack = false
                                }
                            }
                            if (matchMethodInfo.returnType != null) {
                                val returnCtClass = ctMethod!!.returnType
                                val returnType = returnCtClass.name
                                //有设置返回类型这一项
                                if (returnType != matchMethodInfo.returnType) {
                                    isBack = false
                                }
                            }
                            //                            logger.error("paramStr="+paramStr+",returnType="+returnType+",matchMethodInfo="+matchMethodInfo);
                            val methodInfo = ctMethod!!.methodInfo
                            val codeAttribute = methodInfo.codeAttribute
                            if (codeAttribute == null) {
                                isBack = false
                            }
                        } catch (ignored: java.lang.Exception) {
                        }
                        if (isBack) {
                            onCallBackMethod.onBackName(
                                MethodRecord(
                                    name,
                                    descriptor,
                                    aopMatchCut.cutClassName
                                )
                            )
                            //                            cacheMethodRecords.add(new MethodRecord(name, descriptor, aopMatchCut.getCutClassName()));
                            if (cutInfoJson) {
                                putCutInfo(
                                    "匹配切面",
                                    slashToDot(
                                        className
                                    ),
                                    aopMatchCut.cutClassName,
                                    CutMethodJson(name, descriptor, false)
                                )
                            }
                        }
                        break
                    }
                }
            }
        }
        val myMethodVisitor = MyMethodVisitor(
            access, name, descriptor, signature, exceptions, MethodRecord(name, descriptor, null)
        )
        methods.add(myMethodVisitor)
        return myMethodVisitor
    }

    var lambdaMethodList = mutableListOf<LambdaMethod>()


    override fun visitEnd() {
        super.visitEnd()

        this.methods.forEach { methodNode ->

            for (tmpNode in methodNode.instructions) {
                if (tmpNode is InvokeDynamicInsnNode) {
                    val desc = tmpNode.desc
                    val descType = Type.getType(desc)
                    val samBaseType = descType.returnType
                    //sam 接口名
                    val samBase = samBaseType.descriptor
                    //sam 方法名
                    val samMethodName = tmpNode.name
                    val bsmArgs = tmpNode.bsmArgs
                    val samMethodType = bsmArgs[0] as Type
                    val methodName = bsmArgs[1] as Handle
                    //sam 实现方法实际参数描述符
//                  Type implMethodType = (Type) bsmArgs[2];
                    val lambdaName = methodName.name
                    val thisClassName = slashToDot(
                        samBase.substring(1).replace(";".toRegex(), "")
                    ).replace("$", ".")
                    val originalClassName =
                        slashToDot(samBase.substring(1).replace(";".toRegex(), ""))
                    //                    logger.error("className="+className+",tmpNode.name="+tmpNode.name+",desc=" + desc + ",samBase=" + samBase + ",samMethodName="
//                            + samMethodName + ",methodName=" + lambdaName+ ",methodDesc=" + methodName.getDesc()+",thisClassName="+thisClassName+
//                            ",getDescriptor"+samMethodType.getDescriptor());
                    val lambdaDesc = methodName.desc
                    val samMethodDesc = samMethodType.descriptor
                    val lambdaMethod = LambdaMethod(
                        samMethodName,
                        samMethodDesc,
                        thisClassName,
                        originalClassName,
                        lambdaName,
                        lambdaDesc
                    )
                    lambdaMethodList.add(lambdaMethod)
                }
            }

            if (lambdaMethodList.size > 0) {
                WovenInfoUtils.aopMatchCuts.forEach { (_: String?, aopMatchCut: AopMatchCut) ->
                    if (AopMatchCut.MatchType.SELF.name != aopMatchCut.matchType && aopMatchCut.methodNames.size == 1) {
                        for ((_, descriptor, clsName, originalClassName, lambdaName, lambdaDesc) in lambdaMethodList) {
                            val isDirectExtends =
                                slashToDotClassName(aopMatchCut.baseClassName) == clsName
                            var isMatch = false
                            if (AopMatchCut.MatchType.DIRECT_EXTENDS.name == aopMatchCut.matchType) {
                                isMatch = isDirectExtends
                            } else if (AopMatchCut.MatchType.EXTENDS.name == aopMatchCut.matchType || AopMatchCut.MatchType.LEAF_EXTENDS.name == aopMatchCut.matchType) {
                                isMatch = isDirectExtends
                                if (!isMatch) {
                                    val parentClsName = aopMatchCut.baseClassName
                                    isMatch = isInstanceof(clsName, slashToDotClassName(parentClsName))
                                }
                            }
                            val aopMatchCutMethodName = aopMatchCut.methodNames[0]
                            val matchMethodInfo = getMethodInfo(aopMatchCutMethodName)
                            if (isMatch && matchMethodInfo != null && name == matchMethodInfo.name) {
                                var isBack = true
                                try {
                                    val classPool = classPool
                                    var ctMethod: CtMethod? = null
                                    if (matchMethodInfo.paramTypes != null || matchMethodInfo.returnType != null) {
                                        val ctClass = classPool!![originalClassName]
                                        ctMethod = getCtMethod(ctClass, name, descriptor)
                                    }
                                    if (matchMethodInfo.paramTypes != null && ctMethod != null) {
                                        val ctClasses = ctMethod.parameterTypes
                                        val paramStr = StringBuilder()
                                        paramStr.append("(")
                                        val length = ctClasses.size
                                        for (i in 0 until length) {
                                            paramStr.append(ctClasses[i].name)
                                            if (i != length - 1) {
                                                paramStr.append(",")
                                            }
                                        }
                                        paramStr.append(")")
                                        //有设置参数类型这一项
                                        if (paramStr.toString() != matchMethodInfo.paramTypes) {
                                            isBack = false
                                        }
                                    }
                                    if (matchMethodInfo.returnType != null && ctMethod != null) {
                                        val returnCtClass = ctMethod.returnType
                                        val returnType = returnCtClass.name
                                        //有设置返回类型这一项
                                        if (returnType != matchMethodInfo.returnType) {
                                            isBack = false
                                        }
                                    }
                                } catch (ignored: java.lang.Exception) {
                                }
                                if (isBack) {
                                    val methodRecord = MethodRecord(
                                        lambdaName,
                                        lambdaDesc,
                                        aopMatchCut.cutClassName
                                    )
                                    onCallBackMethod.onBackName(methodRecord)
                                    if (cutInfoJson) {
//                                        InitConfig.INSTANCE.putCutInfo("匹配切面",lambdaMethod.getOriginalClassName()+"_"+lambdaMethod.getLambdaName(),aopMatchCut.getCutClassName(),new CutMethodJson(name,returnType == null?(descriptor.substring(0,descriptor.lastIndexOf(")")+1)):paramStr.toString(),returnType == null?(descriptor.substring(descriptor.lastIndexOf(")")+1)):returnType,true));
                                        putCutInfo(
                                            "匹配切面",
                                            originalClassName + "_" + lambdaName,
                                            aopMatchCut.cutClassName,
                                            CutMethodJson(name, descriptor, true)
                                        )
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }


    interface OnCallBackMethod {
        fun onBackName(methodRecord: MethodRecord)
    }
}